smart_card_connector_app/src/chrome-api-jstocxxtest.js

```diff
@@ -640,6 +659,139 @@ goog.exportSymbol('testChromeApiProviderToCpp', {
             'onDisconnectRequested', /*requestId=*/ 113, BAD_HANDLE,
             chrome.smartCardProviderPrivate.Disposition.LEAVE_CARD)
         .$waitAndVerify();
+  },
+
+  'testConnectedReader': {
+    'setUp': async function() {
+      launchPcscServer(/*initialDevices=*/[{
+        'id': 123,
+        'type': SimulationConstants.GEMALTO_DEVICE_TYPE,
+        'cardType': SimulationConstants.COSMO_CARD_TYPE
+      }]);
+      createChromeApiProvider();
+      const sCardContext = await establishContext(/*requestId=*/ 1111);
+      readerHandle = await connect(
+          /*requestId=*/ 1112, sCardContext,
+          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+          chrome.smartCardProviderPrivate.ShareMode.SHARED,
+          /*preferredProtocols=*/ {'t1': true},
+          /*expectedProtocol=*/ chrome.smartCardProviderPrivate.Protocol.T1);
+    },
+
+    'tearDown': async function() {
+      expectReportPlainResult(/*requestId=*/ 2222, 'SUCCESS');
+      await mockChromeApi
+          .dispatchEvent(
+              'onDisconnectRequested', /*requestId=*/ 2222, readerHandle,
+              chrome.smartCardProviderPrivate.Disposition.LEAVE_CARD)
+          .$waitAndVerify();
+    },
+
+    // Test that Status returns information about the card.
+    'testStatus': async function() {
+      expectReportStatusResult(
+          /*requestId=*/ 111,
+          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+          chrome.smartCardProviderPrivate.ConnectionState.NEGOTIABLE,
+          chrome.smartCardProviderPrivate.Protocol.T1,
+          SimulationConstants.COSMO_ID_70_ATR, 'SUCCESS');
+      await mockChromeApi
+          .dispatchEvent('onStatusRequested', /*requestId=*/ 111, readerHandle)
+          .$waitAndVerify();
+    },
+
+    // Test that Status returns READER_UNAVAILABLE after reader disappears.
+    'testStatus_readerUnavailable': async function() {
+      setSimulatedDevices(/*devices=*/[]);
+
+      // Status should eventually return a READER_UNAVAILABLE error.
+      for (let requestId = 111; ; requestId += 1) {
+        const saveArgs = new goog.testing.mockmatchers.SaveArgument();
+        expectReportStatusResult(
+            requestId, saveArgs, saveArgs, saveArgs, saveArgs, saveArgs);
```

Comment:
Does anything guarantee the left-to-right ordering of the argument processing? I'm worried `saveArgs` might turn out to be populated in unexpected order.

Reply:
changed this to saving with `$does`.